File: QuickSort.java
//Quicsort
class QuickSort
{
    int partition(int[] inp,int ub,int lb)
    {
        int start = lb;
        //System.out.println(start);
        int end = ub;
        //System.out.println(end);
        int pivot = inp[start];
        while(start < end)
        {
            while(pivot >= inp[start] && start < ub)
            {
                //System.out.println(start);
                start++;
            }
            while(pivot < inp[end])
            {
                end--;
            }
            //System.out.println("at start:"+inp[start]+" ");
            //System.out.println("at End :"+inp[end]+" ");
            if(start < end)
            {
                swap(inp,start,end);
            }
            
            display(inp);
            }
        System.out.print("swap : ");
        swap(inp,lb,end);
        display(inp);
        return end;
    }
    void swap(int[] arr,int a,int b)
    {
        int temp = arr[a];
        arr[a] = arr[b];
        arr[b] = temp;
        //System.out.println("swap completed!!");
    }
    void display(int[] arr)
    {
        for (int i = 0;i<arr.length;i++)
            System.out.print(arr[i]+" ");
        System.out.println();
    }
    void Qsort(int[] inp,int ub,int lb)
    {
        int loc = 0;
        if(lb < ub)
        {
            //System.out.println("lb and ub "+lb+" "+ub);
            loc  = partition(inp,ub,lb);
            //System.out.println("\n"+loc);
            Qsort(inp,(loc-1),lb); 
            Qsort(inp,ub,(loc+1));
        }
    }
    public static void main(String[] args) {
        int inp[] = {9,7,4,1,3,5,2,8};
        for (int i = 0;i<inp.length;i++)
            System.out.print(inp[i]+" ");
        System.out.println();
        int ub = (inp.length)-1;
        int lb = 0;
        QuickSort qs = new QuickSort();
        qs.Qsort(inp,ub,lb);
    }
}

File: quicksort.js
// QUICK SORT

// STEP 1 : INITIALIZE variables Loc , inputArray[],UpperBound = 0 and LowerBound = inputArray.length-1

// STEP 2 : Calling QuickSort Function

// STEP 3 : Below is QuickSort function

// Function QuickSort(inputArray,LowerBound,UpperBound)
// If condition LowerBound < UpperBound
//      Loc = Partition(inputArray,LowerBound,UpperBound)
//      QuickSort(inputArray,LowerBound,Loc-1)
//      QuickSort(inputArray,Loc+1,UpperBound)

// Step 4 : QuickSort function recursively calls the partion function and returns the previous pivot 

// Function Partition(inputArray,LowerBound,UpperBound)
//      pivot = inputArray[LowerBound]
//      Start = LowerBound
//      End = Uppperbound
// Outer While loop condition Start < End
//     Inner While loop condition pivot >= inputArray[Start] AND Start < UpperBound
//          Start++
//     [end inner while loop]
//     Inner while loop condition pivot > inputArray[end]
//          End++
//     [end inner while loop]
//     If condition Start < End
//          Swapping inputArray[Start] with inputArray[End]
//     [end if]
// [end outer while loop]
// swapping with inputArray[lowerbound] with inputArray[End]
// Return End

// STEP 5 : Quick Sort Function recursively calls the partion function until the List will be sorted

// STEP 6 : End of Quick sort Function result will be in same array/list

/*** BELOW JAVASCRIPT CODE FOR IMPLEMENTATION OF QUICKSORT***/

function partition(inpArr,lowerbound,uppperbound)
{
    let pivot = inpArr[lowerbound];
    let start = lowerbound;
    let end = uppperbound;
    while(start < end)
    {
        while(pivot >= inpArr[start] && start < uppperbound)
            start++;
        while(pivot < inpArr[end])
            end--;
        if(start < end)
            swap(inpArr,start,end);
    }
    swap(inpArr,lowerbound,end);
    return end; 
}
function swap(inpArr,a,b)
{
    let temp = inpArr[a];
    inpArr[a] = inpArr[b];
    inpArr[b] = temp
}
function qsort(inpArr,lowerbound,uppperbound)
{
    let loc = 0;
    if(lowerbound < uppperbound)
    {
        loc = partition(inpArr,lowerbound,uppperbound);
        qsort(inpArr,lowerbound,(loc-1));
        qsort(inpArr,(loc+1),uppperbound);
    }
}
let inpArr = [8,3,4,2,6,7,8,9];
let uppperbound,lowerbound;
lowerbound = 0;
uppperbound = inpArr.length-1;
console.log(inpArr);
qsort(inpArr,lowerbound,uppperbound);
console.log(inpArr);